src/attr.ts

```diff
@@ -0,0 +1,92 @@
+import {CustomElement} from './custom-element'
+
+const attrs = new WeakMap<Record<PropertyKey, unknown>, string[]>()
+type attrValue = string | number | boolean
+
+/**
+ * Attr is a decorator which tags a property as one to be initialized via
+ * `initializeAttrs`.
+ *
+ * The signature is typed such that the property must be one of a String,
+ * Number or Boolean. This matches the behavior of `initializeAttrs`.
+ */
+export function attr<K extends string>(proto: Record<K, attrValue>, key: K): void {
+  if (!attrs.has(proto)) attrs.set(proto, [])
+  attrs.get(proto)!.push(key)
+}
+
+/**
+ * initializeAttrs is called with a set of class property names (if omitted, it
+ * will look for any properties tagged with the `@attr` decorator). With this
+ * list it defines property descriptors for each property that map to `data-*`
+ * attributes on the HTMLElement instance.
+ *
+ * It works around Native Class Property semantics - which are equivalent to
+ * calling `Object.defineProperty` on the instance upon creation, but before
+ * `constructor()` is called.
+ *
+ * If a class property is assigned to the class body, it will infer the type
+ * (using `typeof`) and define an appropriate getter/setter combo that aligns
+ * to that type. This means class properties assigned to Numbers can only ever
+ * be Numbers, assigned to Booleans can only ever be Booleans, and assigned to
+ * Strings can only ever be Strings.
+ *
+ * This is automatically called as part of `@controller`. If a class uses the
+ * `@controller` decorator it should not call this manually.
+ */
+export function initializeAttrs(instance: HTMLElement, names?: Iterable<string>): void {
+  if (!names) names = attrs.get(Object.getPrototypeOf(instance)) || []
+  for (const key of names) {
+    const value = (<Record<PropertyKey, unknown>>(<unknown>instance))[key]
+    const name = attrToAttributeName(key)
+    let descriptor: PropertyDescriptor = {
+      get(this: HTMLElement): string {
+        return String(this.getAttribute(name) || '')
+      },
+      set(this: HTMLElement, newValue: string) {
+        this.setAttribute(name, newValue || '')
+      }
+    }
+    if (typeof value === 'number') {
+      descriptor = {
+        get(this: HTMLElement): number {
+          return Number(this.getAttribute(name) || 0)
+        },
+        set(this: HTMLElement, newValue: string) {
```

Comment:
```suggestion
        set(this: HTMLElement, newValue: number) {
```

Reply:
Semi intentional ;)

This avoids having to call `String(newValue)` or otherwise forcing a cast on L56 because TypeScript doesn't understand that `setAttribute` can receive a value and call `toString()` on it.